test/extensions/http/header_validators/envoy_default/header_validator_test.cc

```diff
@@ -16,80 +10,322 @@ namespace HeaderValidators {
+  }
 
-constexpr absl::string_view empty_config = R"EOF(
-    name: envoy.http.header_validators.envoy_default
-    typed_config:
-        "@type": type.googleapis.com/envoy.extensions.http.header_validators.envoy_default.v3.HeaderValidatorConfig
-)EOF";
+  HeaderEntryValidationResult validateResponseHeaderEntry(const HeaderString&,
+                                                          const HeaderString&) override {
+    return HeaderEntryValidationResult::success();
+  }
 
-} // namespace
+  RequestHeaderMapValidationResult validateRequestHeaderMap(RequestHeaderMap&) override {
+    return RequestHeaderMapValidationResult::success();
+  }
 
-class HeaderValidatorTest : public testing::Test {
+  ResponseHeaderMapValidationResult validateResponseHeaderMap(ResponseHeaderMap&) override {
+    return ResponseHeaderMapValidationResult::success();
+  }
+};
+
+using BaseHttpHeaderValidatorPtr = std::unique_ptr<BaseHttpHeaderValidator>;
+
+class BaseHeaderValidatorTest : public HeaderValidatorTest {
 protected:
-  ::Envoy::Http::HeaderValidatorPtr create(absl::string_view config_yaml,
-                                           ::Envoy::Http::Protocol protocol) {
-    auto* factory =
-        Registry::FactoryRegistry<Envoy::Http::HeaderValidatorFactoryConfig>::getFactory(
-            "envoy.http.header_validators.envoy_default");
-    ASSERT(factory != nullptr);
-
-    envoy::config::core::v3::TypedExtensionConfig typed_config;
+  BaseHttpHeaderValidatorPtr createBase(absl::string_view config_yaml) {
+    envoy::extensions::http::header_validators::envoy_default::v3::HeaderValidatorConfig
+        typed_config;
     TestUtility::loadFromYaml(std::string(config_yaml), typed_config);
 
-    uhv_factory_ = factory->createFromProto(typed_config.typed_config(), context_);
-    return uhv_factory_->create(protocol, stream_info_);
+    return std::make_unique<BaseHttpHeaderValidator>(typed_config, Protocol::Http11, stream_info_);
   }
+};
 
-  void setHeaderStringUnvalidated(HeaderString& header_string, absl::string_view value) {
-    header_string.setCopyUnvalidatedForTestOnly(value);
-  }
+TEST_F(BaseHeaderValidatorTest, ValidateMethodPermissive) {
+  HeaderString valid{"GET"};
+  HeaderString valid_lowercase{"post"};
+  HeaderString custom{"Custom-Method"};
+  auto uhv = createBase(empty_config);
+  EXPECT_TRUE(uhv->validateMethodHeader(valid).ok());
+  EXPECT_TRUE(uhv->validateMethodHeader(valid_lowercase).ok());
+  EXPECT_TRUE(uhv->validateMethodHeader(custom).ok());
+}
 
-  NiceMock<Server::Configuration::MockFactoryContext> context_;
-  ::Envoy::Http::HeaderValidatorFactoryPtr uhv_factory_;
-  NiceMock<Envoy::StreamInfo::MockStreamInfo> stream_info_;
-};
+TEST_F(BaseHeaderValidatorTest, ValidateMethodRestricted) {
+  HeaderString valid{"GET"};
+  HeaderString post_lowercase{"post"};
+  HeaderString custom{"CUSTOM-METHOD"};
+  auto uhv = createBase(restrict_http_methods_config);
+  EXPECT_TRUE(uhv->validateMethodHeader(valid).ok());
```

Comment:
Optional: define macro EXPECT_ACCEPT() that prints details() in case of failure.